bpf/tests/nodeport_hybrid_dsr_test.c

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+/* Copyright Authors of Cilium */
+
+#include <bpf/ctx/skb.h>
+#include "common.h"
+
+#include <linux/types.h>
+
+union v4addr {
+	__be32 addr;
+	__u8 raw[4];
+};
+
+union v6addr {
+	struct {
+		__be32 p1;
+		__be32 p2;
+		__be32 p3;
+		__be32 p4;
+	};
+	__u8 addr[16];
+};
+
+union macaddr {
+	struct {
+		__u32 p1;
+		__u16 p2;
+	};
+	__u8 addr[6];
+};
+
+#include <bpf/config/node.h>
+
+#define ENABLE_DSR 1
+#define ENABLE_DSR_BYUSER 1
+#define ENABLE_IPV4 1
+#define ENABLE_IPV6 1
+
+#define DSR_ENCAP_MODE 1
+#define DSR_ENCAP_IPIP 1
+
+#define SVC_FLAG_ROUTABLE		(1 << 0)
+#define SVC_FLAG_FWD_MODE_DSR		(1 << 7)
+
+struct lb4_service {
+	__u32 backend_id;
+	__u32 count;
+	__u16 rev_nat_index;
+	__u8 flags;
+	__u8 flags2;
+};
+
+struct lb6_service {
+	__u32 backend_id;
+	__u32 count;
+	__u16 rev_nat_index;
+	__u8 flags;
+	__u8 flags2;
+};
+
+static __always_inline bool nodeport_uses_dsr(bool flip __maybe_unused)
```

Comment:
Rather than re-implementing the structs & code, this should include `lib/nodeport.h` to access the actual implementations :).

(if that ends up needing too much boilerplate definitions to make `nodeport.h` compile-able on its own, an easy cheat would be to include the `bpf_host.h` test header and test in context of `bpf_host`)